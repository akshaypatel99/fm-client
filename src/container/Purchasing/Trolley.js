import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link, useHistory, useLocation, useParams } from 'react-router-dom';
import Message from '../../components/Message';
import {
	Button,
	Card,
	Col,
	FormControl,
	Image,
	ListGroup,
	ListGroupItem,
	Row,
} from 'react-bootstrap';
import {
	addToTrolley,
	removeFromTrolley,
} from '../../redux/actions/trolleyActions';

const Trolley = () => {
	const { prodId } = useParams();
	let history = useHistory();
	let location = useLocation();

	const qty = location.search ? Number(location.search.split('=')[1]) : 1;

	const dispatch = useDispatch();

	const trolley = useSelector((state) => state.trolley);
	const { trolleyItems } = trolley;

	const userLogin = useSelector((state) => state.userLogin);
	const { userInfo } = userLogin;

	useEffect(() => {
		if (prodId) {
			dispatch(addToTrolley(prodId, qty));
		}
	}, [dispatch, prodId, qty]);

	const removeFromTrolleyHandler = (id) => {
		dispatch(removeFromTrolley(id));
	};

	const checkoutHandler = () => {
		if (!userInfo) {
			history.push('/login');
		} else {
			history.push('/delivery-address');
		}
	};

	return (
		<>
			<Link className='btn btn-primary my-3' to='/'>
				Return to Shop
			</Link>
			<Row>
				<Col md={8}>
					<h1>Your Shopping Trolley</h1>
					{trolleyItems.length === 0 ? (
						<Message variant='secondary'>
							Your trolley is empty. Return to{' '}
							<Link to='/'>
								<strong>Shop</strong>
							</Link>{' '}
							to add items.
						</Message>
					) : (
						<ListGroup variant='flush'>
							{trolleyItems.map((item) => (
								<ListGroupItem key={item.productId}>
									<Row>
										<Col md={2}>
											<Image src={item.image} alt={item.name} fluid rounded />
										</Col>
										<Col md={3}>
											<Link to={`/product/${item.productId}`}>
												<strong>{item.name}</strong>
											</Link>
										</Col>
										<Col md={2}>
											<strong>£ {item.price}</strong>
										</Col>
										<Col md={2}>
											<FormControl
												as='select'
												value={item.qty}
												onChange={(e) =>
													dispatch(
														addToTrolley(item.productId, Number(e.target.value))
													)
												}
											>
												{[...Array(item.countInStock).keys()]
													.slice(0, 10)
													.map((n) => (
														<option key={n + 1} value={n + 1}>
															{n + 1}
														</option>
													))}
											</FormControl>
										</Col>
										<Col md={2}>
											<Button
												type='button'
												variant='danger'
												onClick={() => removeFromTrolleyHandler(item.productId)}
											>
												<i className='fas fa-trash'></i>
											</Button>
										</Col>
									</Row>
								</ListGroupItem>
							))}
						</ListGroup>
					)}
				</Col>
				<Col md={4}>
					<Card>
						<ListGroupItem variant='flush'>
							<h2>
								Subtotal (
								{trolleyItems.reduce((acc, item) => acc + item.qty, 0)}) items
							</h2>
							£{' '}
							{trolleyItems
								.reduce((acc, item) => acc + item.qty * item.price, 0)
								.toFixed(2)}
						</ListGroupItem>
						<ListGroupItem>
							<Button
								type='button'
								className='btn-block'
								disabled={trolleyItems.length === 0}
								onClick={checkoutHandler}
							>
								Proceed To Checkout
							</Button>
						</ListGroupItem>
					</Card>
				</Col>
			</Row>
		</>
	);
};

export default Trolley;
