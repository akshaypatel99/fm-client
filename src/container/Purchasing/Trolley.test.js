import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { screen } from '@testing-library/react';
import configureStore from 'redux-mock-store';

import Trolley from './Trolley';

const mockStore = configureStore([]);

describe('My Connected React-Redux Component', () => {
	let store;
	let component;

	beforeEach(() => {
		store = mockStore({
			trolley: {
				trolleyItems: [
					{
						productId: '5fa1943bcbfa945cfbf5717b',
						name: 'Chicken',
						image: '/images/chicken.jpg',
						price: 8.79,
						countInStock: 35,
						qty: 1,
					},
				],
			},
		});

		store.dispatch = jest.fn();

		component = renderer.create(
			<Provider store={store}>
				<BrowserRouter>
					<Route path='/' render={(props) => <Trolley {...props} />} />
				</BrowserRouter>
			</Provider>
		);
	});

	it('should render with given state from Redux store', () => {
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('should dispatch an action on button click', () => {
		renderer.act(() => {
			component.root.findAllByType('button')[0].props.onClick();
		});

		expect(store.dispatch).toHaveBeenCalledTimes(1);
		expect(screen.queryByText(/Chicken/)).toBeNull();
		expect(screen.queryByRole('img')).toBeNull();
	});
});
