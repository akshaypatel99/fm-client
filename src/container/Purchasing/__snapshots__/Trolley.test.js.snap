// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Connected React-Redux Component should render with given state from Redux store 1`] = `
Array [
  <a
    className="btn btn-light my-3"
    href="/"
    onClick={[Function]}
  >
    Return to Shop
  </a>,
  <div
    className="row"
  >
    <div
      className="col-md-8"
    >
      <h1>
        Your Shopping Trolley
      </h1>
      <div
        className="list-group list-group-flush"
        onKeyDown={[Function]}
      >
        <div
          className="list-group-item"
          data-rb-event-key={null}
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="row"
          >
            <div
              className="col-md-2"
            >
              <img
                alt="Chicken"
                className="img-fluid rounded"
                src="/images/chicken.jpg"
              />
            </div>
            <div
              className="col-md-3"
            >
              <a
                href="/product/5fa1943bcbfa945cfbf5717b"
                onClick={[Function]}
              >
                <strong>
                  Chicken
                </strong>
              </a>
            </div>
            <div
              className="col-md-2"
            >
              <strong>
                £ 
                8.79
              </strong>
            </div>
            <div
              className="col-md-2"
            >
              <select
                className="form-control"
                onChange={[Function]}
                value={1}
              >
                <option
                  value={1}
                >
                  1
                </option>
                <option
                  value={2}
                >
                  2
                </option>
                <option
                  value={3}
                >
                  3
                </option>
                <option
                  value={4}
                >
                  4
                </option>
                <option
                  value={5}
                >
                  5
                </option>
                <option
                  value={6}
                >
                  6
                </option>
                <option
                  value={7}
                >
                  7
                </option>
                <option
                  value={8}
                >
                  8
                </option>
                <option
                  value={9}
                >
                  9
                </option>
                <option
                  value={10}
                >
                  10
                </option>
              </select>
            </div>
            <div
              className="col-md-2"
            >
              <button
                className="btn btn-danger"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                <i
                  className="fas fa-trash"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-md-4"
    >
      <div
        className="card"
      >
        <div
          className="list-group-item list-group-item-flush"
          disabled={false}
          onClick={[Function]}
        >
          <h2>
            Subtotal (
            1
            ) items
          </h2>
          £
           
          8.79
        </div>
        <div
          className="list-group-item"
          disabled={false}
          onClick={[Function]}
        >
          <button
            className="btn-block btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Proceed To Checkout
          </button>
        </div>
      </div>
    </div>
  </div>,
]
`;
